db/infinispan/src/main/java/org/commonjava/indy/infinispan/data/InfinispanStoreDataManager.java

```diff
@@ -112,49 +114,69 @@ private ArtifactStore readValueByJson( String json, StoreKey key )
     @Override
     protected ArtifactStore removeArtifactStoreInternal( StoreKey key )
     {
-        String json = stores.executeCache( ( c ) -> c.remove( key ) );
-        return readValueByJson( json, key );
+        ArtifactStore removed;
+        String json;
+        synchronized ( inMemoryStores )
+        {
+            removed = inMemoryStores.remove( key );
+            json = stores.executeCache( ( c ) -> c.remove( key ) );
+        }
+        if ( removed == null )
+        {
+            removed = readValueByJson( json, key );
+        }
+        return removed;
     }
 
     @Override
     public void clear( final ChangeSummary summary ) throws IndyDataException
     {
+        inMemoryStores.clear();
         stores.executeCache( c -> {
             c.clear();
             return null;
         } );
     }
 
     @Override
-    public Set<ArtifactStore> getAllArtifactStores() throws IndyDataException
-    {
-        return stores.executeCache( c -> {
-            Set<ArtifactStore> ret = new HashSet<>();
-            c.forEach( ( k, v ) -> {
-                ArtifactStore store = readValueByJson( v, k );
-                if ( store != null )
-                {
-                    ret.add( store );
-                }
+    public Set<ArtifactStore> getAllArtifactStores()
```

Comment:
Agree. But seems it is not easy to do now as this api is used in may other modules now. We might think about how to avoid this type of "full scan" ops in other modules in future.

Reply:
We need flexible queries on the store cache. Maybe we should use RDBMS. In the future, we may have Cassandra for storage and some simple cache like NFC/content-index, RDBMS for store-data, ISPN for ?...not sure... The main problems with ISPN are 1. hard to cluster, 2. hard to query.